cache.go

```diff
@@ -89,96 +89,75 @@ func (c *Cache) Clear() {
-//  * If timeout_max_wait: log error and pretend file doesn't exist
+//  1. Check cache for secret.
+//			* If entry is fresh, return cache entry.
+//			* If entry is not fresh, call backend.
+//  2. Ask backend for secret (with timeout).
+//			* If backend returns success: update cache, return.
+//			* If backend returns deleted: set delayed deletion, return data from cache.
+//  3. If timeout backend deadline hit return whatever we have.
 func (c *Cache) Secret(name string) (*Secret, bool) {
-	failureDeadline := time.After(c.timeouts.MaxWait)
-	var backendDeadline <-chan time.Time // inactive, until backend request starts
+	// Perform cache lookup first
+	cacheResult := c.cacheSecret(name)
 
-	var cachedSecret *Secret
-	resultFromCache := func() (*Secret, bool) {
-		success := cachedSecret != nil
-		return cachedSecret, success
-	}
+	var secret *Secret
+	var success bool
 
-	cacheDone := c.cacheSecret(name)
-	var backendDone chan secretResult
+	if cacheResult != nil {
+		secret = &cacheResult.Secret
+		success = true
+	}
 
-	for {
-		select {
-		case s := <-backendDone:
-			backendDone = nil
-			if s.err == nil {
-				// Always return successful value from backend
-				return s.secret, true
-			}
-			if _, ok := s.err.(SecretDeleted); ok {
-				c.secretMap.Delete(name)
-			}
-			// Backend failed and cache lookup already finished
-			if cacheDone == nil {
-				return resultFromCache()
-			}
-		case s := <-cacheDone:
-			cacheDone = nil
-			if s != nil {
-				cachedSecret = &s.Secret
+	// If cache succeeded, and entry is very recent, return cache result
+	if success && time.Since(cacheResult.Time) < c.timeouts.Fresh {
+		return &cacheResult.Secret, success
+	}
 
-				// If cache entry very recent, return cache result
-				if time.Since(s.Time) < c.timeouts.Fresh {
-					return resultFromCache()
-				}
-			}
+	backendDeadline := time.After(c.timeouts.BackendDeadline)
+	backendDone := c.backendSecret(name)
 
-			// Start backend request and wait until optimistic deadline
-			backendDone = c.backendSecret(name)
-			backendDeadline = time.After(c.timeouts.BackendDeadline)
-		case <-backendDeadline:
-			if cachedSecret != nil {
-				return cachedSecret, true
-			}
-		case <-failureDeadline:
-			c.Errorf("Cache and backend timeout: %v", name)
-			return nil, false
+	select {
+	case s := <-backendDone:
+		if s.err == nil {
+			secret = s.secret
+			success = true
 		}
+		if _, ok := s.err.(SecretDeleted); ok {
+			c.secretMap.Delete(name)
+		}
+	case <-backendDeadline:
+		c.Errorf("Backend timeout on secret fetch for '%s'", name)
 	}
+
+	return secret, success
 }
 
 // SecretList returns a listing of Secrets from cache or a server.
 //
 // Cache logic:
-//  * Ask backend w/ timeout
-//  * If backend returns fast: update cache, return
-//  * If timeout_backend_deadline: return cache entries, background update cache when
-//    backend returns
-//  * If timeout_max_wait: log error and pretend no files
+//  * Fetch cache secret list.
```

Comment:
I feel it's more clear in this case to remove this line. The fetch is implicit in `return cache entries` below.